model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/updater/authsession/RootAuthenticationSessionUpdater.java

```diff
@@ -0,0 +1,241 @@
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+public class RootAuthenticationSessionUpdater extends BaseUpdater<String, RootAuthenticationSessionEntity> implements RootAuthenticationSessionModel {
+
+    private final static Comparator<Map.Entry<String, AuthenticationSessionEntity>> TIMESTAMP_COMPARATOR =
+            Comparator.comparingInt(e -> e.getValue().getTimestamp());
+
+    private final List<Consumer<RootAuthenticationSessionEntity>> changes;
+
+    private RealmModel realm;
+    private KeycloakSession session;
+
+
+    private int authSessionsLimit;
+    private RootAuthenticationSessionUpdater(String key, RootAuthenticationSessionEntity entity, long version, UpdaterState initialState) {
+        super(key, entity, version, initialState);
+        if (entity == null) {
+            assert initialState == UpdaterState.DELETED;
+            changes = List.of();
+            return;
+        }
+        changes = new ArrayList<>(4);
+    }
+
+    public synchronized void initialize(KeycloakSession session, RealmModel realm, int authSessionsLimit) {
+        this.session = session;
+        this.realm = realm;
+        this.authSessionsLimit = authSessionsLimit;
+    }
+
+    /**
+     * @return {@code true} if it is already initialized.
+     */
+    public synchronized boolean isInitialized() {
+        return session != null;
+    }
+
+
+    @Override
+    protected boolean isUnchanged() {
+        return changes.isEmpty();
+    }
+
+    public static RootAuthenticationSessionUpdater create(String key, RootAuthenticationSessionEntity entity) {
+        return new RootAuthenticationSessionUpdater(key, Objects.requireNonNull(entity), NO_VERSION, UpdaterState.CREATED);
+    }
+
+    public static RootAuthenticationSessionUpdater wrap(String key, RootAuthenticationSessionEntity value, long version) {
+        return new RootAuthenticationSessionUpdater(key, Objects.requireNonNull(value), version, UpdaterState.READ);
+    }
+
+    public static RootAuthenticationSessionUpdater delete(String key) {
+        return new RootAuthenticationSessionUpdater(key, null, NO_VERSION, UpdaterState.DELETED);
+    }
+
+    @Override
+    public Expiration computeExpiration() {
+        return new Expiration(
+                SessionTimeouts.getAuthSessionLifespanMS(realm, null, getValue()),
+                SessionTimeouts.getAuthSessionMaxIdleMS(realm, null, getValue()));
+    }
+
+    @Override
+    public RootAuthenticationSessionEntity apply(String ignored, RootAuthenticationSessionEntity cachedEntity) {
+        assert !isDeleted();
+        assert !isReadOnly();
+        if (cachedEntity == null) {
+            //entity removed
+            return null;
+        }
+        changes.forEach(c -> c.accept(cachedEntity));
+        return cachedEntity;
```

Comment:
@graziang For multi-site deployment, this should do the trick to remove the entry from the remote cache when the map is empty. 
```suggestion
        return cachedEntity.getAuthenticationSessions().isEmpty() ? null : cachedEntity;
```

Reply:
Yeah it works, We would need something similar for `RootAuthenticationSessionAdapter`.